### PR TITLE
Thermostat component

### PR DESCRIPTION
**Pull request by kbx81**
## Description:

Same code as #1061, renamed to `thermostat`.

**Related issue (if applicable):** #1061 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#665

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/thermostat/thermostat_climate.cpp [hunk 1]
```diff
@@ -29,6 +29,7 @@ void ThermostatClimate::setup() {
   this->setup_complete_ = true;
   this->publish_state();
 }
+float ThermostatClimate::hysteresis() { return this->hysteresis_; }
```
Comment (ssieb) [inline esphome/components/thermostat/thermostat_climate.cpp:32]: Should probably be named "get_hysteresis()".

Comment (kbx81) [reply esphome/components/thermostat/thermostat_climate.cpp:32]: sgtm! Do we have a standard for naming these types of thing? I'm only aware of:
https://esphome.io/guides/contributing.html#codebase-standards
...and:
https://google.github.io/styleguide/cppguide.html#Function_Names
...the later of which calls out what I did here. I'm happy to do either. 😄

Comment (ssieb) [reply esphome/components/thermostat/thermostat_climate.cpp:32]: Never mind.  I see it does match the other values.

**Comment by dseven** [comment]
Running [my prototype](https://github.com/dseven/esphome_thermostat) (don't laugh too hard) with this now. Seems good so-far.

File: tests/test3.yaml [hunk 2]
```diff
@@ -610,7 +610,7 @@ interval:
         position: !lambda 'return 50/100.0;'
 
 climate:
-  - platform: bang_bang
+  - platform: thermostat
```
Comment (glmnet) [inline tests/test3.yaml:613]: this, as a change, will cause bang_bang to not being tested anymore

Comment (kbx81) [reply tests/test3.yaml:613]: Yes, this is an issue, and all of the constants will need to be added back to `const.py`. I will fix later today. 😄

Comment (glmnet) [reply tests/test3.yaml:613]: Ok, bang_bang is reverted so you can work on top of that

Comment (kbx81) [reply tests/test3.yaml:613]: Perfect! Thank you @glmnet 🍻

**Comment by blademckain** [comment]
more or less when it will be included in the dev version of esphome?

**Comment by ssieb** [comment]
@blademckain it already is.

**Pull request merged**